tokio/src/sync/once_cell.rs

```diff
@@ -157,179 +231,161 @@ impl<T> OnceCell<T> {
+            // permit, then we can set the value.
             match self.semaphore.try_acquire() {
-                Ok(_permit) => {
-                    if !self.initialized() {
-                        // SAFETY: There is only one permit on the semaphore, hence only one
-                        // mutable reference is created
-                        unsafe { self.set_value(value) };
-
-                        return Ok(());
-                    } else {
-                        unreachable!(
-                            "acquired the permit after OnceCell value was already initialized."
-                        );
-                    }
+                Ok(permit) => {
+                    debug_assert!(!self.initialized());
+                    self.set_value(value, permit);
+                    Ok(())
                 }
-                _ => {
-                    // Couldn't acquire the permit, look if initializing process is already completed
-                    if !self.initialized() {
-                        return Err(SetError::InitializingError(value));
-                    }
+                Err(TryAcquireError::NoPermits) => {
+                    // Some other task is holding the permit. That task is
+                    // currently trying to initialize the value.
+                    Err(SetError::InitializingError(value))
+                }
+                Err(TryAcquireError::Closed) => {
+                    // The semaphore was closed. Some other task has initialized
+                    // the value.
+                    Err(SetError::AlreadyInitializedError(value))
                 }
             }
+        } else {
+            Err(SetError::AlreadyInitializedError(value))
         }
-
-        Err(SetError::AlreadyInitializedError(value))
     }
 
-    /// Tries to initialize the value of the OnceCell using the async function `f`.
-    /// If the value of the OnceCell was already initialized prior to this call,
-    /// a reference to that initialized value is returned. If some other thread
-    /// initiated the initialization prior to this call and the initialization
-    /// hasn't completed, this call waits until the initialization is finished.
+    /// Get the value currently in the `OnceCell`, or initialize it with the
+    /// given asynchronous operation.
+    ///
+    /// If some other task is currently working on initializing the `OnceCell`,
+    /// this call will wait for that other task to finish, then return the value
+    /// that the other task produced.
+    ///
+    /// If the provided operation is cancelled or panics, the initialization
+    /// attempt is cancelled. If there are other tasks waiting for the value to
+    /// be initialized, one of them will start another attempt at initializing
+    /// the value.
     ///
-    /// This will deadlock if `f` tries to initialize the cell itself.
+    /// This will deadlock if `f` tries to initialize the cell recursively.
```

Comment:
could be nice to have an example here, but not a blocker.

Reply:
Well, then I would want a normal example too. We can't have the only example be how to do it incorrectly.